File: Programmers/level2/findBigsquare/find_bigsquare.py
# 문제 출처 : https://programmers.co.kr/learn/courses/30/lessons/12905
# 참고 블로그 : https://velog.io/@tjdud0123/가장-큰-정사각형
# DP(Dynamic programming, 동적 계획법) 유형 문제

from itertools import chain


def solution(board):
    m, n = len(board), len(board[0])
    for i in range(1, m):
        for j in range(1, n):
            if board[i][j] == 1:
                board[i][j] = min(board[i-1][j-1], board[i-1]
                                  [j], board[i][j-1]) + 1
    # print(board)
    # print(list(chain(*board)))
    return max(chain(*board))**2

File: Programmers/level2/findBigsquare/findBigsquare.js
// https://programmers.co.kr/learn/courses/30/lessons/12905?language=javascript

export const solution = (board) => {

    let m = board.length;
    let n = board[0].length;
    
    for(let i = 1; i < m; i++){
        for(let j = 1; j < n; j++){
            if(board[i][j] === 1){
               board[i][j] = Math.min(board[i-1][j-1], board[i-1][j], board[i][j-1]) + 1 
            }
        }
    }
    console.log(board)
    let max = 0;
    board.forEach(el => {
        if(Math.max(...el) > max){
            max = Math.max(...el); 
        }
    })

    return max**2;
}
